sale_block_cancellation/models/sale.py

```diff
@@ -0,0 +1,108 @@
+#
+#    Copyright (c) 2016 Vauxoo - http://www.vauxoo.com
+#    All Rights Reserved.
+#    info Vauxoo (info@vauxoo.com)
+#    coded by: Luis Torres <luis_t@vauxoo.com>
+#    planned by: Sabrina Romero <sabrina@vauxoo.com>
+############################################################################
+from openerp import models, fields, api, _
+from openerp.exceptions import ValidationError
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    @api.multi
+    @api.depends('picking_ids', 'picking_ids.state')
+    def _get_allow_cancel(self):
+        '''If all the pickings related to a SO are on "Ready to transfer" or
+        in "Waiting another move" state, the cancellation of this must be
+        allowed, or even whit pickings on different states, if the moved
+        product quantities are correctly returned.'''
+        for sale in self:
+            allow = True
+            for pick in sale.picking_ids:
+                if pick.state not in (
+                        'assigned', 'waiting', 'confirmed', 'done'):
+                    allow = False
+            if not allow or not sale.check_prod_by_locations():
+                sale.allow_cancel = False
+            else:
+                sale.allow_cancel = True
+
+    @api.multi
+    def _get_picking_ids(self):
+        res = super(SaleOrder, self)._get_picking_ids(
+            name='picking_ids', args=None)
+        for sale in self:
+            sale.picking_ids = res.get(sale.id)
+
+    def _search_pickings(self, operator, value):
+        res = []
+        pick_obj = self.env["stock.picking"]
+        pick_ids = []
+        if operator == 'in' and value:
+            pick_ids = pick_obj.browse(value)
+        elif operator and value:
+            pick_ids = pick_obj.search([('name', operator, value)])
+        else:
+            pick_ids = pick_obj.search([])
+        if pick_ids:
+            for pick in pick_ids:
+                proc_gp = pick.group_id
+                if proc_gp:
+                    sale_ids = self.search([
+                        ('procurement_group_id', '=', proc_gp.id)])
+                    res.extend(sale_ids.ids)
+        if operator == '=' and not value:
+            return [('id', 'not in', res)]
+        return [('id', 'in', res)]
+
+    allow_cancel = fields.Boolean(
+        'Allow Cancel', compute=_get_allow_cancel, store=True,
+        help='This field check if all the pickings related to '
+        'this SO are on "Ready to transfer", "Waiting Availability" or in '
+        '"Waiting another move" or if the pickings on different states have '
+        'the move of product quantities correctly returned. If this '
+        'validations are True the sale order cancellation must be allowed.')
+    picking_ids = fields.One2many(
+        compute=_get_picking_ids, method=True, relation='stock.picking',
+        string='Picking associated to this sale', search=_search_pickings)
+
+    @api.multi
+    def check_prod_by_locations(self):
```

Comment:
@luistorresm 

What happen is the warehouse is in three steps to delivery and the customer make a return and return is of customer to stock directly , this work as well?